win32ss/gdi/ntgdi/freetype.c

```diff
@@ -945,6 +1055,155 @@ SubstituteFontRecurse(LOGFONTW* pLogFont)
     return TRUE;    /* success */
 }
 
+static inline VOID
+FontLink_Chain_Init(PFONTLINK_CHAIN pChain, PTEXTOBJ pTextObj, FT_Face face, PDC dc)
+{
+    RtlZeroMemory(pChain, sizeof(*pChain));
+    pChain->pBaseTextObj = pTextObj;
+    pChain->pDefFace = face;
+    pChain->dc = dc;
+}
+
+static VOID
+FontLink_Chain_Free(PFONTLINK_CHAIN pChain)
+{
+    PLIST_ENTRY Entry;
+    PFONTLINK pLink;
+
+    if (!pChain->szzFontLink[0])
+        return;
+
+    while (!IsListEmpty(&pChain->FontLinkList))
+    {
+        Entry = RemoveHeadList(&pChain->FontLinkList);
+        pLink = CONTAINING_RECORD(Entry, FONTLINK, ListEntry);
+        FontLink_Destroy(pLink);
+    }
+}
+
+static NTSTATUS
+FontLink_Chain_LoadReg(PFONTLINK_CHAIN pChain, LPLOGFONTW pLF)
+{
+    NTSTATUS Status;
+    HKEY hKey;
+    DWORD cbData;
+    static const WCHAR s_szzDefFontLink[MAX_FONTLINK] =
+        L"tahoma.ttf,Tahoma\0"
+        L"msgothic.ttc,MS UI Gothic\0"
+        L"mingliu.ttc,PMingLiU\0"
+        L"simsun.ttc,SimSun\0"
+        L"gulim.ttc,Gulim\0"
+        L"\0";
+    static const WCHAR s_szzDefFixedPitchFontLink[MAX_FONTLINK] =
```

Comment:
Does Windows' own win32k.sys (or maybe one of its font driver dlls) do the same thing of hardcoding some default font links?

Reply:
This is just my imagination.